### PR TITLE
Add CI builder for windows-arm64.

### PR DESCRIPTION
**Pull request by pbo-linaro**
This PR adds support for windows-arm64 artifacts, using new infrastructure that will be used from January 2023, and replacing https://flutter.googlesource.com/recipes/.

I checked locally that targets are correctly built and expected zips are there 👍.

All packages can be cross compiled from an x64 machine.
Unittests are disabled, as they require an arm64 machine.

Solves this issue: https://github.com/flutter/flutter/issues/115983

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by pbo-linaro** [comment]
@godofredoc @loic-sharma This is the PR for adding windows-arm64 artifacts.

File: ci/builders/windows_arm_host_engine.json
Comment (godofredoc) [inline ci/builders/windows_arm_host_engine.json:1]: This is great news that no new targets were needed. Can you also add a block to the .ci.yaml to enable this builder?

e.g. https://github.com/flutter/engine/blob/main/.ci.yaml#L499-L504, you'll need to set `bringup: true`

Comment (pbo-linaro) [reply ci/builders/windows_arm_host_engine.json:1]: Done!

**Review by godofredoc — APPROVED** [review]

**Comment by godofredoc** [comment]
@loic-sharma can you please help us with a second review to be able to land this change?

**Comment by pbo-linaro** [comment]
@godofredoc Once this PR gets merged, will artifacts be available within hours, or should we wait until January?

**Comment by godofredoc** [comment]
> @godofredoc Once this PR gets merged, will artifacts be available within hours, or later in January?

This change is safe to land but we need a second review. The artifacts will be generated as soon as this PR is landed and uploaded to a different bucket.

After coming from vacation I can help with the engine.py change for you to get access to the archives from within the flutter sdk.

**Comment by pbo-linaro** [comment]
Is the engine.py modification still needed? From what I understood, this will be deprecated in January and replaced by current approach. But maybe I missed something. The final goal is to add support to [flutter.bat](https://github.com/flutter/flutter).

File: ci/builders/windows_arm_host_engine.json
Comment (loic-sharma) [inline ci/builders/windows_arm_host_engine.json:92]: When I tested this on my x64 machine I got this archive instead: `out/host_release_arm64/zip_archives/windows-x64-release/windows-x64-flutter.zip`

I suspect the output path needs to be updated to use the `$full_target_platform_name` instead: https://github.com/flutter/engine/blob/6e6a6538f9e217b0a5fb94550cf16d7c12ec2c3d/build/archives/BUILD.gn#L227-L230


Here are the commands I used to build:

```
.\tools\gn --runtime-mode release --no-lto --no-goma --windows-cpu arm64
autoninja -C ..\out\host_release_arm64 windows gen_snapshot flutter/build/archives:windows_flutter
```

Comment (pbo-linaro) [reply ci/builders/windows_arm_host_engine.json:92]: I missed that one (windows-arm64 vs windows-arm64-{debug,release}), it's now fixed.
Not sure if target_cpu is always defined for x64 build, but we'll see that with CI.

**Comment by pbo-linaro** [comment]
Fix name of archive generated to reflect target architecture.

**Comment by pbo-linaro** [comment]
@godofredoc Is there something to add somewhere to get arm_host_engine workfow enabled in github CI? That would help to ensure everything is working as expected before merging this.

File: ci/builders/windows_arm_host_engine.json
Comment (loic-sharma) [inline ci/builders/windows_arm_host_engine.json:15]: I got the following archives instead:

1. `out/host_debug_arm64/zip_archives/windows-x64/windows-x64-embedder.zip`
2. `out/host_debug_arm64/zip_archives/windows-x64/flutter-cpp-client-wrapper.zip`

Here are the commands I used to build:

```
.\tools\gn --runtime-mode debug --full-dart-sdk --no-lto --no-goma --windows-cpu arm64
```

```
autoninja -C ..\out\host_debug_arm64 flutter/build/archives:artifacts flutter/build/archives:embedder flutter/tools/font-subset flutter/build/archives:dart_sdk_archive flutter/shell/platform/windows/client_wrapper:client_wrapper_archive flutter/build/archives:windows_flutter flutter/web_sdk
```

Comment (loic-sharma) [reply ci/builders/windows_arm_host_engine.json:15]: It looks like you'll need to update these lines as well:

https://github.com/flutter/engine/blob/1d2ba73d1059544564e644055d4366956e581e87/shell/platform/windows/client_wrapper/BUILD.gn#L116

https://github.com/flutter/engine/blob/1d2ba73d1059544564e644055d4366956e581e87/shell/platform/embedder/BUILD.gn#L533

Comment (pbo-linaro) [reply ci/builders/windows_arm_host_engine.json:15]: Done

**Comment by pbo-linaro** [comment]
I ran a full build to check locally all zips are correct:
```
python3 ./flutter/tools/gn --runtime-mode debug --full-dart-sdk --no-lto --no-goma --windows-cpu arm64
python3 ./flutter/tools/gn --runtime-mode profile --no-goma --windows-cpu arm64
python3 ./flutter/tools/gn --runtime-mode release --no-goma --windows-cpu arm64
ninja -C out/host_debug_arm64 flutter/build/archives:artifacts flutter/build/archives:embedder flutter/tools/font-subset flutter/build/archives:dart_sdk_archive flutter/shell/platform/windows/client_wrapper:client_wrapper_archive flutter/build/archives:windows_flutter flutter/web_sdk || exit 1
ninja -C out/host_profile_arm64 windows gen_snapshot flutter/build/archives:windows_flutter
ninja -C out/host_release_arm64 windows gen_snapshot flutter/build/archives:windows_flutter
```

And I get this list of zip files:
```
./src/out/host_debug_arm64/zip_archives/dart-sdk-windows-arm64.zip
./src/out/host_debug_arm64/zip_archives/flutter-web-sdk-windows-arm64.zip
./src/out/host_debug_arm64/zip_archives/windows-arm64/artifacts.zip
./src/out/host_debug_arm64/zip_archives/windows-arm64/flutter-cpp-client-wrapper.zip
./src/out/host_debug_arm64/zip_archives/windows-arm64/font-subset.zip
./src/out/host_debug_arm64/zip_archives/windows-arm64/windows-arm64-embedder.zip
./src/out/host_debug_arm64/zip_archives/windows-arm64-debug/windows-arm64-flutter.zip
./src/out/host_profile_arm64/zip_archives/windows-arm64-profile/windows-arm64-flutter.zip
./src/out/host_release_arm64/zip_archives/windows-arm64-release/windows-arm64-flutter.zip
```

I should have check everything manually, sorry. I was expecting CI to do its job for me on this.

**Comment by chinmaygarde** [comment]
From Triage: Please wait for @zanderso s review on this one.

**Review by zanderso — APPROVED** [review]
lgtm

**Review by loic-sharma — APPROVED** [review]
LGTM! Nice work 🥳

**Comment by pbo-linaro** [comment]
Thanks for your precious help @loic-sharma!

**Comment by loic-sharma** [comment]
@pbo-linaro This change had to be reverted as it broke the `Mac mac_host_engine` build with:

```
Step('Copy gs://flutter_archives_v2/flutter_infra_release/flutter/5a39a8846621408403e5ccb3b078f2e4327ae1eb/flutter-web-sdk-darwin-x64.zip') failed 'ENOENT' with: [Errno 2] No such file or directory: '/Volumes/Work/s/w/ir/cache/builder/src/out/host_debug/zip_archives/flutter-web-sdk-darwin-x64.zip'
```

My guess is this change caused a mismatch between:

https://github.com/flutter/engine/blob/264aa032cf754f6a010ecb52f248a4df648bb0c6/web_sdk/BUILD.gn#L554

And:

https://github.com/flutter/engine/blob/264aa032cf754f6a010ecb52f248a4df648bb0c6/ci/builders/mac_host_engine.json#L12

**Comment by pbo-linaro** [comment]
Probably adding an if for windows platform is the easiest thing to do.

Just a question: What does the whole CI check regarding artifacts?
I expected missing or wrongly named packages to be found *before* anything gets merged in trunk.

Are we sure it's the only package that was broken?

**Comment by pbo-linaro** [comment]
Created a new PR adding a special case for web-sdk artifacts: https://github.com/flutter/engine/pull/38739

**Pull request merged**
